File: src/main/java/pl/training/workshop/m8/s08_compilergate/step3/SeatMap.java
package pl.training.workshop.m8.s08_compilergate.step3;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/** Krok 3 (bez zmian): mapa miejsc na typach generycznych. */
public final class SeatMap {
    private final Map<Integer, List<String>> seatsByRow = new TreeMap<>();

    public void take(String seat) {
        Integer row = Integer.valueOf(seat.substring(1));
        seatsByRow.computeIfAbsent(row, r -> new ArrayList<>()).add(seat);
    }

    public Map<Integer, Integer> takenPerRow() {
        Map<Integer, Integer> result = new TreeMap<>();
        seatsByRow.forEach((row, seats) -> result.put(row, seats.size()));
        return result;
    }
}

File: typescript/src/workshop/m8/s08_compilergate/start/SeatMap.ts
/**
 * Start: mapa zajętych miejsc na "surowym" typie - new Map() bez argumentów typu to Map<any, any>
 * (rawtypes) - i niesprawdzanych asercjach typu na wartościach any (unchecked).
 */
export class SeatMap {
  private readonly seatsByRow = new Map();

  take(seat: string): void {
    const row = Number.parseInt(seat.substring(1), 10);
    let seats = this.seatsByRow.get(row) as string[] | undefined;
    if (seats === undefined) {
      seats = [];
      this.seatsByRow.set(row, seats);
    }
    seats.push(seat);
  }

  takenPerRow(): Map<number, number> {
    const result = new Map<number, number>();
    for (const row of [...this.seatsByRow.keys()].sort((a, b) => a - b)) {
      result.set(row as number, (this.seatsByRow.get(row) as string[]).length);
    }
    return result;
  }
}
